File: apps/agent/buildplan_test.go
package main

import (
	"os"
	"path/filepath"
	"strings"
	"testing"
)

func TestPrepareDockerfilePrefersExistingDockerfile(t *testing.T) {
	dir := t.TempDir()
	if err := os.WriteFile(filepath.Join(dir, "Dockerfile"), []byte("FROM scratch\n"), 0o600); err != nil {
		t.Fatal(err)
	}
	path, plan, err := prepareDockerfile(dir, "")
	if err != nil {
		t.Fatal(err)
	}
	if path != "Dockerfile" || plan != "dockerfile:auto" {
		t.Fatalf("unexpected plan %q %q", path, plan)
	}
}

func TestPrepareDockerfileGeneratesNode24Plan(t *testing.T) {
	dir := t.TempDir()
	manifest := `{"scripts":{"build":"vite build","start":"tsx server/index.ts"},"engines":{"node":">=24"}}`
	if err := os.WriteFile(filepath.Join(dir, "package.json"), []byte(manifest), 0o600); err != nil {
		t.Fatal(err)
	}
	if err := os.WriteFile(filepath.Join(dir, "package-lock.json"), []byte("{}"), 0o600); err != nil {
		t.Fatal(err)
	}
	path, plan, err := prepareDockerfile(dir, "")
	if err != nil {
		t.Fatal(err)
	}
	if plan != "nodejs-24:auto" {
		t.Fatalf("unexpected plan %q", plan)
	}
	contents, err := os.ReadFile(filepath.Join(dir, path))
	if err != nil {
		t.Fatal(err)
	}
	text := string(contents)
	for _, expected := range []string{"FROM node:24-bookworm-slim", "npm ci --no-audit --no-fund", "RUN npm run build", `CMD ["npm","start"]`} {
		if !strings.Contains(text, expected) {
			t.Fatalf("generated Dockerfile missing %q:\n%s", expected, text)
		}
	}
}

func TestPrepareDockerfileRequiresStartScript(t *testing.T) {
	dir := t.TempDir()
	if err := os.WriteFile(filepath.Join(dir, "package.json"), []byte(`{"scripts":{"build":"npm run compile"}}`), 0o600); err != nil {
		t.Fatal(err)
	}
	if _, _, err := prepareDockerfile(dir, ""); err == nil {
		t.Fatal("expected missing start script to fail")
	}
}

File: apps/builder/src/build-plan.test.ts
import assert from "node:assert/strict";
import test from "node:test";
import { mkdtemp, readFile, rm, writeFile } from "node:fs/promises";
import { tmpdir } from "node:os";
import { join } from "node:path";
import { node24BaseImage, prepareBuildDockerfile } from "./build-plan";

test("builder prefers existing Dockerfile", async () => {
  const dir = await mkdtemp(join(tmpdir(), "rundea-builder-plan-"));
  try {
    await writeFile(join(dir, "Dockerfile"), "FROM scratch\n");
    assert.deepEqual(await prepareBuildDockerfile(dir, null), { dockerfile:"Dockerfile", plan:"dockerfile:auto" });
  } finally {
    await rm(dir,{recursive:true,force:true});
  }
});

test("builder generates pinned Node plan", async () => {
  const dir = await mkdtemp(join(tmpdir(), "rundea-builder-plan-"));
  try {
    await writeFile(join(dir, "package.json"), JSON.stringify({scripts:{build:"vite build",start:"node server.js"}}));
    await writeFile(join(dir, "package-lock.json"), "{}");
    const result = await prepareBuildDockerfile(dir, null);
    assert.equal(result.plan, "nodejs-24.20.0:auto");
    const text = await readFile(join(dir,result.dockerfile),"utf8");
    assert.equal(text.includes("FROM " + node24BaseImage), true);
    assert.equal(text.includes("npm ci --no-audit --no-fund"), true);
    assert.equal(text.includes("RUN npm run build"), true);
    assert.equal(text.includes('CMD ["npm","start"]'), true);
  } finally {
    await rm(dir,{recursive:true,force:true});
  }
});

test("builder Node plan requires scripts.start", async () => {
  const dir = await mkdtemp(join(tmpdir(), "rundea-builder-plan-"));
  try {
    await writeFile(join(dir, "package.json"), JSON.stringify({scripts:{build:"echo build"}}));
    await assert.rejects(() => prepareBuildDockerfile(dir,null), /scripts.start/);
  } finally {
    await rm(dir,{recursive:true,force:true});
  }
});
